src/sentry/auth/staff.py

```diff
@@ -0,0 +1,319 @@
+from __future__ import annotations
+
+import ipaddress
+import logging
+from datetime import datetime, timedelta
+
+from django.conf import settings
+from django.core.signing import BadSignature
+from django.utils import timezone
+from django.utils.crypto import constant_time_compare, get_random_string
+from rest_framework.request import Request
+
+from sentry.auth.elevated_mode import ElevatedMode
+from sentry.auth.system import is_system_auth
+from sentry.utils.auth import has_completed_sso
+
+logger = logging.getLogger("sentry.staff")
+
+SESSION_KEY = "_stf"
+
+COOKIE_NAME = getattr(settings, "STAFF_COOKIE_NAME", "stf")
+
+COOKIE_SALT = getattr(settings, "STAFF_COOKIE_SALT", "")
+
+COOKIE_SECURE = getattr(settings, "STAFF_COOKIE_SECURE", settings.SESSION_COOKIE_SECURE)
+
+COOKIE_DOMAIN = getattr(settings, "STAFF_COOKIE_DOMAIN", settings.SESSION_COOKIE_DOMAIN)
+
+COOKIE_PATH = getattr(settings, "STAFF_COOKIE_PATH", settings.SESSION_COOKIE_PATH)
+
+COOKIE_HTTPONLY = getattr(settings, "STAFF_COOKIE_HTTPONLY", True)
+
+# the maximum time the session can stay alive
+MAX_STAFF_SESSION_AGE = timedelta(hours=4)
+
+IDLE_MAX_STAFF_SESSION_AGE = timedelta(minutes=15)
+
+ALLOWED_IPS = frozenset(getattr(settings, "STAFF_ALLOWED_IPS", settings.INTERNAL_IPS) or ())
+
+ORG_ID = getattr(settings, "STAFF_ORG_ID", None)
+
+DISABLE_SSO_CHECK_FOR_LOCAL_DEV = getattr(settings, "DISABLE_SSO_CHECK_FOR_LOCAL_DEV", False)
+
+UNSET = object()
+
+
+def is_active_staff(request: Request) -> bool:
+    if is_system_auth(getattr(request, "auth", None)):
+        return True
+    stf = getattr(request, "staff", None) or Staff(request)
+    return stf.is_active
+
+
+class Staff(ElevatedMode):
+    allowed_ips = frozenset(ipaddress.ip_network(str(v), strict=False) for v in ALLOWED_IPS)
+
+    def __init__(self, request, allowed_ips=UNSET, current_datetime=None):
+        self.uid: str | None = None
+        self.request = request
+        if allowed_ips is not UNSET:
+            self.allowed_ips = frozenset(
+                ipaddress.ip_network(str(v), strict=False) for v in allowed_ips or ()
+            )
+        self._populate(current_datetime=current_datetime)
+
+    @property
+    def is_active(self):
+        # We have a wsgi request with no user.
+        if not hasattr(self.request, "user"):
+            return False
+        # if we've been logged out
+        if not self.request.user.is_authenticated:
+            return False
+        # if staff status was changed
+        if not self.request.user.is_staff:
+            return False
+        # if the user has changed
+        if str(self.request.user.id) != self.uid:
+            return False
+        return self._is_active
```

Comment:
After reading the code all the way, I now know `_populate` is at the end supposed to help establish this value before being referenced, but if that's the case, it would be better in the `__init__` to establish this value first as a default for safer code. Reading this code from top to bottom, it's not clear that the value gets dynamically added to the instance; I thought "Oh no, we're referencing a value that doesn't exist!" haha

We have to go 2 levels deep (from populate -> set_logged_in/out) to know this is a class variable we can use in this class, and if something happens in that chain, we could be blowing up here/throwing an error.